### PR TITLE
CASMPET-4302 Main: KeycloakPlugin: Update documentation and configure Nexus

### PR DESCRIPTION
**Pull request by ndavidson-hpe**
## Summary and Scope

Updates the documentation on how to authenticate to nexus via the keycloak plugin. This is a slight change in function to how logging into nexus worked before. The only changes are to how to log into nexus with keycloak accounts.

## Issues and Related PRs

* Resolves [CASMPET-4302](https://jira-pro.its.hpecorp.net:8443/browse/CASMPET-4302)

### CODE REVIEW

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 1]
```diff
@@ -10,11 +10,28 @@ Use the hostname set in `istio.ingress.hosts.ui.authority` \(see below\) to conn
 https://nexus.{{network.dns.external}}/
 ```
 
-Users will be redirected to Keycloak to log in, and based on the default OPA policy, only admin users will be authorized. Scripts may connect by first obtaining a JWT token from Keycloak and passing it in the HTTP `Authorization` header.
+### Using Keycloak to Create and Manage Accounts
+
+To log into the Web UI or authenticate with the REST API a user account with approate permissions must be created. Accounts are managed in keycloak \(see keycloak account documentation [here](../CSM_product_management/Configure_Keycloak_Account.md)\). To add administrator permissions for nexus add the nx-admin role binding to the user from the system-nexus-client client \(see below\). To add an anonymous user add the nx-anonymous role binding to the user from the sysyem-nexus-client client \(see below\).
```
Comment (kimjensen-hpe) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:15]: typo sysyem-nexus-client

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 2]
```diff
@@ -10,11 +10,28 @@ Use the hostname set in `istio.ingress.hosts.ui.authority` \(see below\) to conn
...
+### Using the Local Nexus Admin Account
+
+During the deployment or update of Nexus a local admin account is created. To access the local admin account for Nexus on any ncn run the following commands:
+
+```bash
+kubectl -n nexus get secret nexus-admin-credential --template {{.data.username}} | base64 -d; echo
+kubectl -n nexus get secret nexus-admin-credential --template {{.data.password}} | base64 -d; echo
+```
+
+The first command will return with username of the local admin account, and the second will return the password for the local admin account. \(Note the secret will not update or stay in sync if the username or password of the local account is changed.\). This account has the same permissions as an account creted in keycloak with the nx-admin role.
```
Comment (kimjensen-hpe) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:30]: typo creted

**Review by kimjensen-hpe — COMMENTED** [review]
A couple of minor typos

**Review by rnoska-hpe — APPROVED** [review]
Approved for merge after typos are fixed.

File: operations/package_repository_management/Repair_Yum_Repository_Metadata.md [hunk 3]
```diff
@@ -22,7 +22,9 @@ The system is fully installed.
     https://nexus.{{network.dns.external}}/
     ```
 
-    Users will be redirected to Keycloak to log in, and based on the default OPA policy, only admin users will be authorized. Scripts may connect by first obtaining a JWT token from Keycloak and passing it in the HTTP Authorization header.
+    Users will need to login through nexus UI. The account is configured through keycloak with a role mapping for nexus authentication. The role needed for admin permissions is nx-admin in the system-nexus-client role. Scripts may connect by using a username and password in the request.
```
Comment (nrockershousen) [inline operations/package_repository_management/Repair_Yum_Repository_Metadata.md:25]: Keycloak and Nexus should be capitalized

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 4]
```diff
@@ -379,5 +397,39 @@ For example:
 # curl -sfkSL -X DELETE "https://packages.local/service/rest/v1/blobstores/NAME"
 ```
 
+### Authenticating to Access the REST API
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:400]: Change "authenticating" to "authenticate"

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 5]
```diff
@@ -379,5 +397,39 @@ For example:
...
+
+To access some of the REST API functions not listed above you need to authenticate using a username and password. This username and password is the same used to sign into the Web UI. Either the username and password of a properly permissioned Keycloak account or the nexus local admin account must be used. To get the nexus local admin account after a fresh install you can run the follow fuction:
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:402]: Minor wording change:

---

An authenticated username and password are required to access some of the REST API functions not listed above. This username and password are the same used to sign into the Web UI. Either the username and password of a properly permissioned Keycloak account or the Nexus local admin account must be used. 

Use the following function to get the nexus local admin account after a fresh install:

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 6]
```diff
@@ -379,5 +397,39 @@ For example:
...
+    [[ $# -gt 0 ]] || set -- -n nexus nexus-admin-credential
+
+    kubectl get secret "${@}" >/dev/null || return $?
+
+    NEXUS_USERNAME="$(kubectl get secret "${@}" --template {{.data.username}} | base64 -d)"
+    NEXUS_PASSWORD="$(kubectl get secret "${@}" --template {{.data.password}} | base64 -d)"
+}
+```
+
+Then to use the REST API you could use the following line \(The second line uses a keycloak account with username:USERNAME and password:PASSWORD this is only an example and they should be replaced with the proper username and password\):
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:425]: Wording/formatting change:

---

Authenticate either the Keycloak or Nexus account to use the REST API.

To authenticate the Nexus local admin account:

```bash
# Nexus-local user
curl -i -sfv -u "$NEXUS_USERNAME:$NEXUS_PASSWORD" -H "accept: application/json" -X GET https://packages.local/service/rest/beta/security/user-sources
```

To authenticate a Keycloak account:

This example uses a Keycloak account with username:USERNAME and password:PASSWORD. Replace these values with the proper username and password before running the command.

```bash
# Keycloak user
curl -i -sfv -u "USERNAME:PASSWORD" -H "accept: application/json" -X GET https://packages.local/service/rest/beta/security/user-sources
```

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 7]
```diff
@@ -10,11 +10,28 @@ Use the hostname set in `istio.ingress.hosts.ui.authority` \(see below\) to conn
...
+kubectl -n nexus get secret nexus-admin-credential --template {{.data.password}} | base64 -d; echo
+```
+
+The first command will return with username of the local admin account, and the second will return the password for the local admin account. \(Note the secret will not update or stay in sync if the username or password of the local account is changed.\). This account has the same permissions as an account created in keycloak with the nx-admin role.
 
 ### Access Nexus with the REST API
 
-The [REST API](https://help.sonatype.com/repomanager3/rest-and-integration-api) is available from NCNs or compute nodes at https://packages.local/service/rest, as well as over the CAN at https://nexus.\{\{network.dns.external\}\}/service/rest \(requires JWT token in the HTTP `Authorization` header\).
+The [REST API](https://help.sonatype.com/repomanager3/rest-and-integration-api) is available from NCNs or compute nodes at https://packages.local/service/rest, as well as over the CAN at https://nexus.\{\{network.dns.external\}\}/service/rest \(requires authentication with username and password\).
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:34]: Change "CAN" to "Customer Access Network (CAN)"

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 8]
```diff
@@ -10,11 +10,28 @@ Use the hostname set in `istio.ingress.hosts.ui.authority` \(see below\) to conn
 https://nexus.{{network.dns.external}}/
 ```
 
-Users will be redirected to Keycloak to log in, and based on the default OPA policy, only admin users will be authorized. Scripts may connect by first obtaining a JWT token from Keycloak and passing it in the HTTP `Authorization` header.
+### Using Keycloak to Create and Manage Accounts
+
+To log into the Web UI or authenticate with the REST API a user account with approate permissions must be created. Accounts are managed in keycloak \(see keycloak account documentation [here](../CSM_product_management/Configure_Keycloak_Account.md)\). To add administrator permissions for nexus add the nx-admin role binding to the user from the system-nexus-client client \(see below\). To add an anonymous user add the nx-anonymous role binding to the user from the system-nexus-client client \(see below\).
+
+    ![Nexus Admin Account](../../img/operations/Nexus_Admin_Account.png "Nexus Admin Account")
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:17]: These image lines shouldn't be indented. They are rendering as a code block.

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 9]
```diff
@@ -10,11 +10,28 @@ Use the hostname set in `istio.ingress.hosts.ui.authority` \(see below\) to conn
 https://nexus.{{network.dns.external}}/
 ```
 
-Users will be redirected to Keycloak to log in, and based on the default OPA policy, only admin users will be authorized. Scripts may connect by first obtaining a JWT token from Keycloak and passing it in the HTTP `Authorization` header.
+### Using Keycloak to Create and Manage Accounts
+
+To log into the Web UI or authenticate with the REST API a user account with approate permissions must be created. Accounts are managed in keycloak \(see keycloak account documentation [here](../CSM_product_management/Configure_Keycloak_Account.md)\). To add administrator permissions for nexus add the nx-admin role binding to the user from the system-nexus-client client \(see below\). To add an anonymous user add the nx-anonymous role binding to the user from the system-nexus-client client \(see below\).
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:15]: These 2 instances of Keycloak should be capitalized.

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 10]
```diff
@@ -10,11 +10,28 @@ Use the hostname set in `istio.ingress.hosts.ui.authority` \(see below\) to conn
 https://nexus.{{network.dns.external}}/
 ```
 
-Users will be redirected to Keycloak to log in, and based on the default OPA policy, only admin users will be authorized. Scripts may connect by first obtaining a JWT token from Keycloak and passing it in the HTTP `Authorization` header.
+### Using Keycloak to Create and Manage Accounts
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:13]: Change "using" to "use"

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 11]
```diff
@@ -10,11 +10,28 @@ Use the hostname set in `istio.ingress.hosts.ui.authority` \(see below\) to conn
...
+
+To log into the Web UI or authenticate with the REST API a user account with approate permissions must be created. Accounts are managed in keycloak \(see keycloak account documentation [here](../CSM_product_management/Configure_Keycloak_Account.md)\). To add administrator permissions for nexus add the nx-admin role binding to the user from the system-nexus-client client \(see below\). To add an anonymous user add the nx-anonymous role binding to the user from the system-nexus-client client \(see below\).
+
+    ![Nexus Admin Account](../../img/operations/Nexus_Admin_Account.png "Nexus Admin Account")
+
+    ![Nexus Anonymous Account](../../img/operations/Nexus_Anonymous_Account.png "Nexus Anonymous Account")
+
+### Using the Local Nexus Admin Account
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:21]: Change "using" to "use"

File: operations/package_repository_management/Manage_Repositories_with_Nexus.md [hunk 12]
```diff
@@ -10,11 +10,28 @@ Use the hostname set in `istio.ingress.hosts.ui.authority` \(see below\) to conn
...
+
+During the deployment or update of Nexus a local admin account is created. To access the local admin account for Nexus on any ncn run the following commands:
+
+```bash
+kubectl -n nexus get secret nexus-admin-credential --template {{.data.username}} | base64 -d; echo
+kubectl -n nexus get secret nexus-admin-credential --template {{.data.password}} | base64 -d; echo
+```
+
+The first command will return with username of the local admin account, and the second will return the password for the local admin account. \(Note the secret will not update or stay in sync if the username or password of the local account is changed.\). This account has the same permissions as an account created in keycloak with the nx-admin role.
```
Comment (nrockershousen) [inline operations/package_repository_management/Manage_Repositories_with_Nexus.md:30]: Keycloak should be capitalized

**Review by phalseth-hpe — APPROVED** [review]

**Pull request merged**